File: 1722-minimize-hamming-distance-after-swap-operations/solution.py
from typing import List
from collections import defaultdict, Counter

class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        
    def find(self, i):
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]
    
    def unite(self, i, j):
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            self.parent[root_i] = root_j

class Solution:
    def minimumHammingDistance(self, source: List[int], target: List[int], allowedSwaps: List[List[int]]) -> int:
        n = len(source)
        uf = UnionFind(n)
        
        # Build components
        for a, b in allowedSwaps:
            uf.unite(a, b)
            
        # Group source elements by component root
        # key: root index, value: Counter of elements in that component
        groups = defaultdict(Counter)
        for i in range(n):
            root = uf.find(i)
            groups[root][source[i]] += 1
            
        matches = 0
        # Calculate how many target elements can be satisfied within each component
        for i in range(n):
            root = uf.find(i)
            val = target[i]
            if groups[root][val] > 0:
                matches += 1
                groups[root][val] -= 1
                
        return n - matches

File: 1722-minimize-hamming-distance-after-swap-operations/solution.cpp
#include <vector>
#include <numeric>
#include <unordered_map>

using namespace std;

class UnionFind {
public:
    vector<int> parent;
    UnionFind(int n) {
        parent.resize(n);
        iota(parent.begin(), parent.end(), 0);
    }
    
    int find(int i) {
        if (parent[i] == i) return i;
        return parent[i] = find(parent[i]);
    }
    
    void unite(int i, int j) {
        int root_i = find(i);
        int root_j = find(j);
        if (root_i != root_j) {
            parent[root_i] = root_j;
        }
    }
};

class Solution {
public:
    int minimumHammingDistance(vector<int>& source, vector<int>& target, vector<vector<int>>& allowedSwaps) {
        int n = source.size();
        UnionFind uf(n);
        
        // Build connected components of indices
        for (const auto& swap : allowedSwaps) {
            uf.unite(swap[0], swap[1]);
        }
        
        // Group source elements by their component root
        unordered_map<int, unordered_map<int, int>> components;
        for (int i = 0; i < n; ++i) {
            int root = uf.find(i);
            components[root][source[i]]++;
        }
        
        int matches = 0;
        // For each index, check if target[i] exists in its component's multiset
        for (int i = 0; i < n; ++i) {
            int root = uf.find(i);
            int targetVal = target[i];
            
            if (components[root].count(targetVal) && components[root][targetVal] > 0) {
                matches++;
                components[root][targetVal]--;
            }
        }
        
        return n - matches;
    }
};